File: sdks/typescript/src/runner.ts
/**
 * Run the `skilltest` CLI as a subprocess and parse its JSON contract.
 *
 * This is the code-level API: call {@link runSkill}, get a typed
 * {@link Report}, assert on `report.passed`, and mix in deterministic checks
 * against the transcript.
 */
import { spawn } from "node:child_process";
import {
  constants,
  accessSync,
  chmodSync,
  existsSync,
  mkdtempSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { createRequire } from "node:module";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { type TestCaseInput, compileCase, isTestCaseInput } from "./case.js";
import {
  SkilltestError,
  SkilltestProviderError,
  SkilltestUsageError,
  providerErrorFor,
} from "./errors.js";
import type { ReportError } from "./generated/error.js";
import type { Report } from "./generated/report.js";
import type { ValidationReport } from "./generated/validation.js";
import { type ToolSpy, bindMocks, compileDecls } from "./mock.js";

/** Environment variables supplying defaults for the binary and provider. */
export const ENV_BIN = "SKILLTEST_BIN";
export const ENV_PROVIDER = "SKILLTEST_PROVIDER";

export interface RunOptions {
  /** Path to the `skilltest` binary (default: `$SKILLTEST_BIN` or `skilltest`). */
  bin?: string;
  /** Provider command (default: `$SKILLTEST_PROVIDER`). A string or argv array. */
  provider?: string | string[];
  /** Harness platforms to run on (overrides config). */
  platforms?: string[];
  /** Models to run on (overrides config). */
  models?: string[];
  /** Model used for evals and the simulated user. */
  judgeModel?: string;
  /** Cap on assistant turns for multi-turn cases. */
  maxTurns?: number;
  /** Path to a config file. */
  config?: string;
  /** Working directory for the subprocess. */
  cwd?: string;
  /**
   * Mock/spy objects ({@link import("./mock.js").spy | spy} /
   * {@link import("./mock.js").stub | stub} /
   * {@link import("./mock.js").deny | deny} /
   * {@link import("./mock.js").rewrite | rewrite}): mocks compile into the
   * run's hook-side ruleset (prepended to the case's own `mocks:` block, so
   * the test-local rule wins), and after the run every object is bound with
   * the calls it matched — assert on it directly. Each run re-binds fresh.
   */
  mocks?: ToolSpy[];
}

interface Captured {
  status: number | null;
  stdout: string;
  stderr: string;
}

const require = createRequire(import.meta.url);

/**
 * The optional platform package that carries the prebuilt binary for this host,
 * e.g. `@skill-test/cli-linux-x64`. One is published per supported target and
 * declared in `optionalDependencies`; the package manager installs only the one
 * matching the host's `os`/`cpu`, so this name resolves to an installed package
 * on exactly the supported platforms.
 */
export function platformPackage(): string {
  return `@skill-test/cli-${process.platform}-${process.arch}`;
}

/**
 * Absolute path to the binary bundled in the matching platform package, or
 * `undefined` when none is installed — a source/dev checkout (the package links
 * but ships no binary), or a platform with no published package. Callers fall
 * back to `$SKILLTEST_BIN`/`PATH`.
 */
export function bundledBin(): string | undefined {
  try {
    const pkgJson = require.resolve(`${platformPackage()}/package.json`);
    const exe = process.platform === "win32" ? "skilltest.exe" : "skilltest";
    const bin = join(dirname(pkgJson), "bin", exe);
    if (!existsSync(bin)) return undefined;
    ensureExecutable(bin);
    return bin;
  } catch {
    return undefined;
  }
}

// Some packers (pnpm pack) drop the executable bit; restore it best-effort. The
// platform packages publish via `npm` (which preserves +x), so this only matters
// as a fallback — and a read-only install keeps the packed mode regardless.
function ensureExecutable(bin: string): void {
  try {
    accessSync(bin, constants.X_OK);
  } catch {
    try {
      chmodSync(bin, 0o755);
    } catch {
      // best effort; if it is not executable and not chmod-able, the spawn fails
      // with a clear EACCES that points at $SKILLTEST_BIN.
    }
  }
}

/**
 * Resolve the binary to run, most explicit first: an explicit `bin`, then
 * `$SKILLTEST_BIN`, then the bundled platform binary, then `skilltest` on PATH.
 */
export function resolveBin(bin: string | undefined): string {
  return bin ?? process.env[ENV_BIN] ?? bundledBin() ?? "skilltest";
}

export function resolveProvider(provider: string | string[] | undefined): string | undefined {
  const value = provider ?? process.env[ENV_PROVIDER];
  if (value === undefined) return undefined;
  return Array.isArray(value) ? value.join(" ") : value;
}

/**
 * The `skilltest run` fragments a `case` becomes (shared with the streaming
 * API): a YAML file/directory rides as a positional path; a code-defined
 * {@link TestCaseInput} is written to a temp JSON file passed as `--case-json`.
 * Call `cleanup()` once the run has finished with the file.
 */
export function caseRunArgs(caseInput: string | TestCaseInput): {
  args: string[];
  cleanup: () => void;
} {
  if (typeof caseInput === "string") return { args: [caseInput], cleanup: () => {} };
  const dir = mkdtempSync(join(tmpdir(), "skilltest-case-"));
  const file = join(dir, "case.json");
  writeFileSync(file, JSON.stringify([compileCase(caseInput)]));
  return {
    args: ["--case-json", file],
    cleanup: () => {
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch {
        // best-effort temp cleanup
      }
    },
  };
}

/**
 * Build the `skilltest run` args for output format `format` (`json` for the
 * buffered API, `json-stream` for the streaming API). `caseArgs` are the
 * fragments identifying the case(s) — a positional path or `--case-json <file>`
 * (see {@link caseRunArgs}). Shared by {@link runSkill} and the streaming API.
 */
export function buildRunArgs(
  caseArgs: string[],
  options: RunOptions,
  format: string,
  mockArgs: string[] = [],
): string[] {
  const args: string[] = [];
  if (options.config) args.push("--config", options.config);
  args.push("run", ...caseArgs, "--format", format);

  const provider = resolveProvider(options.provider);
  if (provider !== undefined) args.push("--provider", provider);
  for (const platform of options.platforms ?? []) args.push("--platform", platform);
  for (const model of options.models ?? []) args.push("--model", model);
  if (options.judgeModel) args.push("--judge-model", options.judgeModel);
  if (options.maxTurns !== undefined) args.push("--max-turns", String(options.maxTurns));
  args.push(...mockArgs);
  return args;
}

/**
 * The CLI flags a `mocks` option turns into (shared with the streaming API):
 * `--spy` so the observation channel is on for spies, plus `--mocks <file>`
 * carrying the compiled mock declarations in a temp dir. Call `cleanup()` once
 * the run has finished with the file.
 */
export function mockRunArgs(mocks: readonly ToolSpy[] | undefined): {
  args: string[];
  cleanup: () => void;
} {
  if (!mocks || mocks.length === 0) return { args: [], cleanup: () => {} };
  const args = ["--spy"];
  const decls = compileDecls(mocks);
  if (decls.length === 0) return { args, cleanup: () => {} };
  const dir = mkdtempSync(join(tmpdir(), "skilltest-mocks-"));
  const file = join(dir, "mocks.json");
  writeFileSync(file, JSON.stringify(decls));
  return {
    args: [...args, "--mocks", file],
    cleanup: () => {
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch {
        // best-effort temp cleanup
      }
    },
  };
}

/**
 * The structured error the CLI emits on stdout for a `--format json` failure, or
 * `undefined` when stdout is not that envelope (an older binary that printed
 * nothing, or a non-JSON line) — callers fall back to the stderr text. Shared by
 * the buffered and streaming APIs. Lightweight-guarded rather than validated:
 * an error envelope has a string `code`/`message` a Report never has.
 */
export function parseReportError(stdout: string): ReportError | undefined {
  const text = stdout.trim();
  if (!text) return undefined;
  try {
    const obj = JSON.parse(text);
    if (
      obj &&
      (obj.code === "usage" || obj.code === "provider") &&
      typeof obj.message === "string"
    ) {
      return obj as ReportError;
    }
  } catch {
    // not JSON — fall back to the stderr text
  }
  return undefined;
}

/**
 * Map a skilltest exit code to a thrown error (shared by the buffered and
 * streaming APIs). Codes 0/1 produce a report and never throw. `structured`,
 * when the CLI emitted the JSON error envelope, carries the classified
 * `kind`/`context` onto a {@link SkilltestProviderError}.
 */
export function raiseForCode(code: number | null, detail: string, structured?: ReportError): void {
  if (code === 0 || code === 1 || code === null) return;
  const message = structured?.message || detail;
  if (code === 2) throw new SkilltestUsageError(message);
  if (code === 3) {
    // The kind-specific subclass (SkilltestTimeoutError, …) so a handler can
    // `instanceof`-check one category; falls back to the base for other/none.
    throw providerErrorFor(message, {
      kind: structured?.kind ?? undefined,
      context: structured?.context ?? undefined,
    });
  }
  throw new SkilltestError(`skilltest exited ${code}: ${message}`);
}

function capture(bin: string, args: string[], cwd: string | undefined): Promise<Captured> {
  return new Promise((resolve, reject) => {
    const child = spawn(bin, args, { cwd });
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (chunk: Buffer) => {
      stdout += chunk.toString();
    });
    child.stderr.on("data", (chunk: Buffer) => {
      stderr += chunk.toString();
    });
    child.on("error", (err) =>
      reject(
        new SkilltestProviderError(
          `could not run skilltest binary \`${bin}\`: ${err.message}. Set ${ENV_BIN} or pass bin.`,
        ),
      ),
    );
    child.on("close", (status) => resolve({ status, stdout, stderr }));
  });
}

// Exit codes that still produce a JSON report (0 = all passed, 1 = some failed).
function raiseForStatus(result: Captured): void {
  if (result.status === 0 || result.status === 1) return;
  const structured = parseReportError(result.stdout);
  const detail = result.stderr.trim() || result.stdout.trim();
  raiseForCode(result.status, detail, structured);
}

// The cast is sound by construction: the SDK's types are generated from the
// CLI's own JSON Schemas and the contract drift gate (`just gen-contract
// --check` in CI) fails when they diverge, so the shape is not re-validated
// here at runtime.
function parse<T>(stdout: string): T {
  try {
    return JSON.parse(stdout) as T;
  } catch (err) {
    throw new SkilltestError(`skilltest did not emit JSON: ${(err as Error).message}`);
  }
}

/**
 * Run one or more test cases and return the parsed {@link Report}. `caseInput`
 * is a code-defined {@link TestCaseInput} (the recommended form) or a path to a
 * YAML file/directory. A failing eval is reported in `report.passed`, not
 * thrown; only bad input ({@link SkilltestUsageError}) and provider failures
 * ({@link SkilltestProviderError}) throw.
 */
export async function runSkill(
  caseInput: string | TestCaseInput,
  options: RunOptions = {},
): Promise<Report> {
  const caseArgs = caseRunArgs(caseInput);
  const mocks = mockRunArgs(options.mocks);
  let result: Captured;
  try {
    const args = buildRunArgs(caseArgs.args, options, "json", mocks.args);
    result = await capture(resolveBin(options.bin), args, options.cwd);
  } finally {
    mocks.cleanup();
    caseArgs.cleanup();
  }
  raiseForStatus(result);
  const report = parse<Report>(result.stdout);
  const caseMocks = isTestCaseInput(caseInput) ? (caseInput.mocks ?? []) : [];
  const bound = [...caseMocks, ...(options.mocks ?? [])];
  if (bound.length > 0) bindMocks(bound, report.runs);
  return report;
}

/** Validate a skill directory (or a folder of them) and return findings. */
export async function validateSkill(
  path: string,
  options: Pick<RunOptions, "bin" | "cwd"> = {},
): Promise<ValidationReport> {
  const args = ["validate", path, "--format", "json"];
  const result = await capture(resolveBin(options.bin), args, options.cwd);
  raiseForStatus(result);
  return parse<ValidationReport>(result.stdout);
}

File: sdks/python/skilltest_sdk/runner.py
"""Run the ``skilltest`` CLI as a subprocess and parse its JSON contract.

This is the code-level API: call [`run_skill`][skilltest_sdk.runner.run_skill],
get a typed [`Report`], assert on ``report.passed``, and mix in any
deterministic checks against the transcript.
"""

from __future__ import annotations

import contextlib
import json
import os
import subprocess
import tempfile
from collections.abc import Iterator, Sequence
from pathlib import Path

from pydantic import BaseModel, ValidationError

from ._error import ReportError
from .case import TestCase
from .errors import SkilltestError, SkilltestProviderError, SkilltestUsageError
from .mock import ToolSpy, bind_mocks, compile_decls
from .models import Report, ValidationReport

#: Environment variables that supply defaults so callers (test-framework
#: packages, CI) can locate the binary and provider without per-call arguments.
ENV_BIN = "SKILLTEST_BIN"
ENV_PROVIDER = "SKILLTEST_PROVIDER"

# Exit codes that still produce a JSON report (0 = all passed, 1 = some failed).
_REPORTING_CODES = frozenset({0, 1})


#: Name of the bundled binary inside the wheel's ``_bin/`` directory.
_BIN_NAME = "skilltest.exe" if os.name == "nt" else "skilltest"


def _bundled_bin() -> str | None:
    """Path to the binary bundled in this wheel, or ``None`` when absent.

    Platform wheels ship the prebuilt CLI at ``skilltest_sdk/_bin/skilltest``;
    the pure (``py3-none-any``) wheel and a source checkout ship none, so callers
    fall back to ``$SKILLTEST_BIN``/``PATH``. Wheel packing can drop the
    executable bit, so restore it best-effort before handing back the path.
    """
    candidate = Path(__file__).resolve().parent / "_bin" / _BIN_NAME
    if not candidate.is_file():
        return None
    if not os.access(candidate, os.X_OK):
        with contextlib.suppress(OSError):
            candidate.chmod(0o755)
    return str(candidate)


def _resolve_bin(bin: str | Path | None) -> str:
    """Resolve the binary, most explicit first: an explicit ``bin``, then
    ``$SKILLTEST_BIN``, then the binary bundled in a platform wheel, then
    ``skilltest`` on ``PATH``."""
    if bin is not None:
        return str(bin)
    env = os.environ.get(ENV_BIN)
    if env:
        return env
    return _bundled_bin() or "skilltest"


def _resolve_provider(provider: str | Sequence[str] | None) -> str | None:
    if provider is None:
        provider = os.environ.get(ENV_PROVIDER)
    if provider is None:
        return None
    if isinstance(provider, str):
        return provider
    return " ".join(provider)


def _run(argv: list[str], cwd: str | Path | None) -> subprocess.CompletedProcess[str]:
    try:
        return subprocess.run(
            argv,
            capture_output=True,
            text=True,
            cwd=cwd,
            check=False,
        )
    except FileNotFoundError as exc:
        raise SkilltestProviderError(
            f"could not run skilltest binary `{argv[0]}`: {exc}. Set {ENV_BIN} or pass bin=..."
        ) from exc


def run_skill(
    case: str | Path | TestCase,
    *,
    bin: str | Path | None = None,
    provider: str | Sequence[str] | None = None,
    platforms: Sequence[str] = (),
    models: Sequence[str] = (),
    judge_model: str | None = None,
    max_turns: int | None = None,
    config: str | Path | None = None,
    cwd: str | Path | None = None,
    mocks: Sequence[ToolSpy] = (),
) -> Report:
    """Run one or more test cases and return the parsed [`Report`].

    ``case`` is a [`TestCase`][skilltest_sdk.case.TestCase] built in code (the
    recommended form), a test-case YAML file, or a directory of them. A failing
    eval is *not* an exception — it is reported in
    ``report.passed``/``report.runs`` so the caller can assert and inspect. Only
    bad input ([`SkilltestUsageError`]) and provider failures
    ([`SkilltestProviderError`]) raise.

    ``mocks`` takes [`spy`][skilltest_sdk.mock.spy] /
    [`stub`][skilltest_sdk.mock.stub] / [`deny`][skilltest_sdk.mock.deny] /
    [`rewrite`][skilltest_sdk.mock.rewrite] objects: mocks compile into the
    run's hook-side ruleset (prepended to the case's own `mocks:` block, so
    the test-local rule wins), and after the run every object is bound with
    the calls it matched — assert on it directly. A ``TestCase``'s own
    ``mocks`` are bound the same way. Each call re-binds fresh.
    """
    case_mocks = tuple(case.mocks) if isinstance(case, TestCase) else ()
    with case_run_args(case) as case_args, mock_run_args(mocks) as mock_args:
        argv = build_run_argv(
            case_args,
            bin=bin,
            provider=provider,
            platforms=platforms,
            models=models,
            judge_model=judge_model,
            max_turns=max_turns,
            config=config,
            fmt="json",
            mock_args=mock_args,
        )
        proc = _run(argv, cwd)
    _raise_for_status(proc)
    report = _parse(Report, proc.stdout)
    bound = [*case_mocks, *mocks]
    if bound:
        bind_mocks(bound, report.runs)
    return report


@contextlib.contextmanager
def case_run_args(case: str | Path | TestCase) -> Iterator[list[str]]:
    """The ``skilltest run`` argv fragments a ``case`` becomes (internal, shared
    with the streaming API): a YAML file/directory rides as a positional path; a
    code-defined [`TestCase`][skilltest_sdk.case.TestCase] is written to a temp
    JSON file passed as ``--case-json``. The temp file lives for the ``with``
    body (the run)."""
    if not isinstance(case, TestCase):
        yield [str(case)]
        return
    fd, path = tempfile.mkstemp(prefix="skilltest-case-", suffix=".json")
    try:
        with os.fdopen(fd, "w") as handle:
            json.dump([case._compile()], handle)
        yield ["--case-json", path]
    finally:
        with contextlib.suppress(OSError):
            os.unlink(path)


@contextlib.contextmanager
def mock_run_args(mocks: Sequence[ToolSpy]) -> Iterator[list[str]]:
    """The CLI flags a ``mocks=`` argument turns into (internal, shared with
    the streaming API): ``--spy`` so the observation channel is on for spies,
    plus ``--mocks <tempfile>`` carrying the compiled mock declarations. The
    temp file lives for the duration of the ``with`` body (the run)."""
    if not mocks:
        yield []
        return
    args = ["--spy"]
    decls = compile_decls(mocks)
    if not decls:
        yield args
        return
    fd, path = tempfile.mkstemp(prefix="skilltest-mocks-", suffix=".json")
    try:
        with os.fdopen(fd, "w") as handle:
            json.dump(decls, handle)
        yield [*args, "--mocks", path]
    finally:
        with contextlib.suppress(OSError):
            os.unlink(path)


def build_run_argv(
    case_args: Sequence[str],
    *,
    bin: str | Path | None,
    provider: str | Sequence[str] | None,
    platforms: Sequence[str],
    models: Sequence[str],
    judge_model: str | None,
    max_turns: int | None,
    config: str | Path | None,
    fmt: str,
    mock_args: Sequence[str] = (),
) -> list[str]:
    """Build the ``skilltest run`` argv for output format ``fmt`` (``json`` for the
    buffered API, ``json-stream`` for the streaming API). ``case_args`` are the
    fragments identifying the case(s) — a positional path or ``--case-json
    <file>`` (see [`case_run_args`][skilltest_sdk.runner.case_run_args]).
    Internal, shared by ``run_skill`` and the streaming API."""
    argv = [_resolve_bin(bin)]
    if config is not None:
        argv += ["--config", str(config)]
    argv += ["run", *case_args, "--format", fmt]

    resolved_provider = _resolve_provider(provider)
    if resolved_provider is not None:
        argv += ["--provider", resolved_provider]
    for platform in platforms:
        argv += ["--platform", platform]
    for model in models:
        argv += ["--model", model]
    if judge_model is not None:
        argv += ["--judge-model", judge_model]
    if max_turns is not None:
        argv += ["--max-turns", str(max_turns)]
    argv += list(mock_args)
    return argv


def validate_skill(
    path: str | Path,
    *,
    bin: str | Path | None = None,
    cwd: str | Path | None = None,
) -> ValidationReport:
    """Validate a skill directory (or a folder of them) and return findings."""
    argv = [_resolve_bin(bin), "validate", str(path), "--format", "json"]
    proc = _run(argv, cwd)
    _raise_for_status(proc)
    return _parse(ValidationReport, proc.stdout)


def _raise_for_status(proc: subprocess.CompletedProcess[str]) -> None:
    if proc.returncode in _REPORTING_CODES:
        return
    structured = parse_report_error(proc.stdout)
    raise_for_code(proc.returncode, proc.stderr.strip() or proc.stdout.strip(), structured)


def parse_report_error(stdout: str) -> ReportError | None:
    """The structured error the CLI emits on stdout for a ``--format json``
    failure, or ``None`` when stdout is not that envelope — an older binary that
    printed nothing to stdout, or a human-format run. Callers fall back to the
    stderr text. Shared by the buffered and streaming APIs."""
    text = stdout.strip()
    if not text:
        return None
    try:
        return ReportError.model_validate_json(text)
    except ValidationError:
        return None


def raise_for_code(
    code: int | None,
    detail: str,
    structured: ReportError | None = None,
) -> None:
    """Map a skilltest exit code to an exception (shared by the buffered and
    streaming APIs). Codes 0/1 produce a report and never raise. ``structured``,
    when the CLI emitted the JSON error envelope, carries the classified
    ``kind``/``context`` onto a [`SkilltestProviderError`]."""
    if code in _REPORTING_CODES:
        return
    message = (structured.message if structured else "") or detail
    if code == 2:
        raise SkilltestUsageError(message)
    if code == 3:
        kind = structured.kind if structured else None
        context = structured.context if structured else None
        raise SkilltestProviderError(message, kind=kind, context=context)
    raise SkilltestError(f"skilltest exited {code}: {message}")


def _parse[T: BaseModel](model: type[T], stdout: str) -> T:
    try:
        return model.model_validate_json(stdout)
    except ValidationError as exc:
        raise SkilltestError(f"skilltest output did not match the expected schema: {exc}") from exc
